File: 26.ListNode链表/leetcode_2130_pairSum.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    int pairSum(ListNode* head) {
        int n = 0;
        ListNode* tmp = head;
        while (tmp){
            n++;
            tmp = tmp->next;
        }

        n /= 2;
        int pos = n;
        ListNode* cur2 = head;
        while (pos){
            cur2 = cur2->next;
            pos--;
        }

        vector<int> res(n);
        ListNode* cur1 = head;
        int indx = 0;
        while (cur2){
            res[indx] += cur1->val;
            res[n-indx-1] += cur2->val;
            cur1 = cur1->next;
            cur2 = cur2->next;
            indx++;
        }
        return *max_element(res.begin(), res.end());
    }
};

int main(){
    Solution s;
    vector<int> nums = {5,4,2,1};
    ListNode* ln = new ListNode();
    ListNode* head = ln;
    for (int i : nums){
        ListNode* cur = new ListNode(i);
        ln->next = cur;
        ln = ln->next;
    }
    cout << s.pairSum(head->next) << endl;
}

File: 26.ListNode链表/leetcode_2130_pairSum.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        n = 0
        tmp = head
        while tmp:
            n += 1
            tmp = tmp.next
        
        n //= 2
        cur2 = head
        pos = n
        while pos:
            cur2 = cur2.next
            pos -= 1
        
        cur1 = head
        res = [0]*n
        indx = 0
        while cur2:
            res[indx] += cur1.val
            res[n-indx-1] += cur2.val
            cur1 = cur1.next
            cur2 = cur2.next
            indx += 1
        return max(res)

if __name__ == "__main__":
    s = Solution()
    nums = [5,4,2,1]
    ln = ListNode()
    head = ln
    for i in nums:
        ln.next = ListNode(i)
        ln = ln.next
    print(s.pairSum(head.next))
